RaftMerge/RaftMerge.tla

```diff
@@ -175,21 +209,128 @@ ClientRequest(i, r, log) ==
+         \* advance apply_index.
+         LET
+           min_index == raft[i][RegionB].logs[next_index].min_index
+           commit_index == next_index
+           fetch_logs == SubSeq(raft[i][RegionB].logs, min_index, commit_index)
+         IN
+           raft' = [InternalRequest(
+                      i, RegionA,
+                      [type         |-> LogMerge,
+                       min_index    |-> min_index,
+                       commit_index |-> commit_index,
+                       entries      |-> fetch_logs]
+                    )
+                    EXCEPT ![i][RegionB].apply_index = next_index]
+       ELSE
+         \* Otherwise, only advance apply_index.
+         raft' = [raft EXCEPT ![i][RegionB].apply_index = next_index]
+    /\ region' = [region EXCEPT ![i][RegionB] = RegionMerging]
+    /\ UNCHANGED <<messages, client_vars>>
+
+\* Apply LogMerge.
+\*
+\* This action is roughly divided into two sub-actions, and executed separately.
+\* The first step copies the logs to region B, to ensure it in sync with leader
+\* B. The second step waits until the copied logs in the first step are applied,
+\* then advances apply_index and marks this region as tombstone.
+ApplyMergeLogStep1(i) ==
+  LET
+    next_index   == raft[i][RegionA].apply_index + 1
+    min_index    == raft[i][RegionA].logs[next_index].min_index
+    commit_index == raft[i][RegionA].logs[next_index].commit_index
+    new_logs     == SubSeq(raft[i][RegionB].logs, 1, min_index - 1)
+                    \o raft[i][RegionA].logs[next_index].entries
+                    \o SubSeq(raft[i][RegionB].logs, commit_index + 1, Len(raft[i][RegionB].logs))
+  IN
+    /\ raft' = [raft EXCEPT ![i][RegionB].logs = new_logs,
+                            ![i][RegionB].commit_index = Max({@, commit_index})]
+    /\ UNCHANGED <<messages, region, client_vars>>
+
+ApplyMergeLogStep2(i) ==
+  LET
+    next_index   == raft[i][RegionA].apply_index + 1
+    commit_index == raft[i][RegionA].logs[next_index].commit_index
+  IN
+    /\ \* Lag logs have been applied.
+       raft[i][RegionB].apply_index >= commit_index
+    /\ raft' = [raft EXCEPT ![i][RegionA].apply_index = next_index]
+    /\ region' = [region EXCEPT ![i][RegionB] = RegionTombStone]
+    /\ UNCHANGED <<messages, client_vars>>
+
+ApplyMergeLog(i) ==
+  LET
+    next_index == raft[i][RegionA].apply_index + 1
+  IN
+    /\ LogAppliable(i, RegionA)
+    /\ raft[i][RegionA].logs[next_index].type = LogMerge
+    /\ \/ ApplyMergeLogStep1(i)
+       \/ ApplyMergeLogStep2(i)
+
+\* Apply LogNormal.
+\* This log simply increases apply_index.
+ApplyNormalLog(i, r) ==
   LET
     next_index == raft[i][r].apply_index + 1
   IN
     /\ LogAppliable(i, r)
-    /\ raft' = [raft EXCEPT ![i][r].apply_index = next_index]
+    /\ raft[i][r].logs[next_index].type = LogNormal
+    /\ LET
+         \* Apply this log if this region is in normal state, otherwise skip it.
+         num_applied_delta == IF region[i][r] = RegionNormal THEN 1 ELSE 0
```

Comment:
if the region is not in the normal mode, it will discard the raft log?

can we allow this?

Reply:
It is a little cheating to discard logs if not in normal mode.

The motivation of writing this is, I don't want to introduce the concept of `epoch` in this code. I think this will both complicate and duplicate code. IIUC, `epoch` is for TiKV to detect whether version has changed or not. If changed, we will skip logs when applying them. Roughly, `epoch` change can be classified into these categories.

1. A premerge Log is applied (region B).
2. Config for region A is changed (which will cause region B to rollback later).

For case 1, isn't it equivalent to not in normal mode?
For case 2, I will add a new constant `WillPerformRollback` to hint that.

These can significantly reduce the state space to explore.

Reply:
I see. For case 1, it is equivalent, because premerge is not normal mode. 

I suggest adding more comments to explain why we can meet this scenario here.